### PR TITLE
Add a lists tutorial

### PR DESCRIPTION
**Pull request by johnwhitington**
Oddly, ocaml.org has never had a tutorial on lists. This PR adds one, and deletes a section from "Data Types and Matching" which is about lists. Next week I will propose a PR to replace the "Data Types and Matching" tutorial too.

(A little content is re-used from "A First Hour with OCaml", but edited for length).

### CODE REVIEW

**Comment by patricoferris** [comment]
Thanks @johnwhitington this is great (as with all of the other tutorial revamping work) :))

File: site/learn/tutorials/lists.md [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+=> 1 + length [2; 3]
+=> 1 + (1 + length [3])
+=> 1 + (1 + (1 + length []))
+=> 1 + (1 + (1 + 0))
+=> 1 + (1 + 1)
+=> 1 + 2
+=> 3
+```
+
+For long lists, this may overflow the stack. The solution is to write our
```
Comment (avsm) [inline site/learn/tutorials/lists.md]: This may be delving into things that are too basic, but we haven't actually introduced what a stack _vs_ heap is, and so tail recursion may come as a surprise this early in the tutorial.

File: site/learn/tutorials/lists.md [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+
+let rev l =
+  List.fold_left (fun a e -> e :: a) [] l;;
+
+let split l =
+  List.fold_right
+    (fun (x, y) (xs, ys) -> (x :: xs, y :: ys))
+    l
+    ([], []);;
+```
```
Comment (avsm) [inline site/learn/tutorials/lists.md:392]: Is it worth linking to https://ocaml.org/learn/tutorials/comparison_of_standard_containers.html for the reader to skip ahead to seeing more data structures like lists if they desire?  The tutorial is nicely linear at the moment though.

Comment (johnwhitington) [reply site/learn/tutorials/lists.md:392]: Done. I don't mind occasional forward references in hypertext, I just don't like them in old-fashioned books.

File: site/learn/tutorials/lists.md (hunk 1 above)
Comment (johnwhitington) [reply site/learn/tutorials/lists.md]: I have moved this section down, even past the Standard Library part, explained in layman's terms what a stack overflow is, and tweaked some language.

**Comment by johnwhitington** [comment]
All done. Thanks.

**Comment by avsm** [comment]
Oh dear, this one conflicts with data_types_and matching.md now. I fear I may have merged in the wrong order @johnwhitington ...

**Comment by johnwhitington** [comment]
I have resolved the conflict. Thanks.

**Comment by avsm** [comment]
My apologies for missing this, merging now. (and cc @patricoferris for a potential sync with v3)

File: site/learn/tutorials/lists.md [hunk 3]
```diff
@@ -0,0 +1,459 @@
+<!-- ((! set title Lists !)) ((! set learn !)) -->
+
+*Table of contents*
+
+#Lists
+
+A list is an ordered sequence of elements. All elements of a list in OCaml must
+be the same type. Lists are built into the language and have a special syntax.
```
Comment (sanette) [inline site/learn/tutorials/lists.md:8]: **of** the same type (?)

File: site/learn/tutorials/lists.md [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+Here is a list of three integers:
+
+```ocamltop
+[1; 2; 3]
+```
+
+Note semicolons separate the elements, not commas. The empty list is written
+`[]`. The type of this list of integers is `int list`.
```
Comment (sanette) [inline site/learn/tutorials/lists.md:16]: I would exchange the order of these sentences, as:
```
The type of this list of integers is `int list`. The empty list is written`[]`. 
```

File: site/learn/tutorials/lists.md [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+```ocamltop
+let rec total l =
+  match l with
+  | [] -> 0
+  | h :: t -> h + total t;;
+
+total [1; 3; 5; 3; 1];;
+```
+
+Consider a function to find the length of a list:
```
Comment (sanette) [inline site/learn/tutorials/lists.md:59]: maybe mention that this function exists (of course) in the standard library

**Comment by johnwhitington** [comment]
@sanette this PR has been merged already. I don't object to any of your changes, though, if you wish to make a new PR.
